Adjust processing version to be integer

vkuznet opened the pull request:
Fixes #10844 

#### Status
not tested

#### Description
Adjust processing version to be integer data type similar to DBS.

#### Is it backward compatible (if not, which system it affects?)
MAYBE

#### Related PRs
#11040 

#### External dependencies / deployment changes
<Does it require deployment changes? Does it rely on third-party libraries?>

vkuznet committed:
Adjust processing version to be integer

vkuznet commented:
@amaltaro , I don't know what's wrong with github, but it does not update ability to merge automatically this PR. the only changes I put is to the schema. The jenkins says `GitHub pull request #11043 of commit 7c588dfc1194ce2a49a9593311559cb1ce0845f7, has merge conflicts.` but I don't see any issue with merge on github. Any clues?

vkuznet commented:
Alan, if possible, can you please create from your end the PR with this changes and we may then close this one. I don't know why github stuck here and I want to check if schema changes will lead to something in jenkins tests.

amaltaro commented:
Let's give it some time. I think GH is going through some changes (I saw new features around) and maybe things didn't go as they expected.

vkuznet commented:
Alan, it looks now things with github get back to normal and now I see that PR can be merged. Could you please trigger the jenkins test.

amaltaro commented:
can you try trigerring it yourself with a comment in this PR as `test this please` (nothing else!). You should have permissions to do so.

vkuznet commented:
test this please

vkuznet commented:
Alan, the good news is that I have no longer see failures in DBS3Upload due to wrong data-type. So, the change to the schema is correct one. The bad news is that I found that in https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/DBS3Buffer/MySQL/UpdateDataset.py#L29 the `processingVer` is passed with a wrong data-type (`None`) which causes failure in unit tests now. I fixed it and resubmit changes, let's see how unit tests will go now. We may have other places where code passes wrong data-type.

vkuznet commented:
Alan, I was trying to converge on minimum set of changes due to wrong data-type and turns out that in order to pass all unit tests I was needed changed both to schema and the ones I introduced in https://github.com/dmwm/WMCore/pull/11040. I merged later over here and now all tests ares passed, and I no longer have issue with WMA (even though now I face a different issue with file parentage which is not related to these changes). So I think we good to go, and I would appreciate if you'll make a final review.

As you said, we may need clean setup of WMA and this can be done once we agree on this PR. If you accept this one we may safely close #11040 as it duplicates the changes now.

vkuznet commented:
With these changes now I see successful data injection into new DBS server:
```
[2022-03-21 08:21:03.455144197 +0000 UTC m=+222376.335625738] logging.go:228: HTTP/1.1 200 POST /dbs2go-writer/bulkblocks [data: 11460 in 2 out] [remoteAddr: 188.184.72.217:47943] [X-Forwarded-For: 188.184.88.70] [X-Forwarded-Host: cmsweb-testbed.cern.ch] [auth: no-TLS cipher-none "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0192.cern.ch" wmagent@vocms0192.cern.ch X509Cert] [ref: "-" "DBSClient/Unknown/"] [req: 412.51213ms proxy-resp: 0]
```
The `bulkblocks` API happily returns 200 HTTP repsonse.

vkuznet commented:
Finally, after many iterations I got completed WMAgent workflow, see https://cmsweb-testbed.cern.ch/reqmgr2/fetch?rid=wmagent_TC_6Tasks_Scratch_test_dbs2go_writer_220322_140254_8786 All bulkblocks API calls to new DBS Go writer are successful
```
[2022-03-22 14:14:57.16858942 +0000 UTC m=+4404.505114946] logging.go:228: HTTP/1.1 200 POST /dbs2go-writer/bulkblocks [data: 13955 in 2 out] [remoteAddr: 10.100.147.128:61518] [X-Forwarded-For: 188.184.88.70] [X-Forwarded-Host: cmsweb-testbed.cern.ch] [auth: no-TLS cipher-none "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0192.cern.ch" wmagent@vocms0192.cern.ch X509Cert] [ref: "-" "DBSClient/Unknown/"] [req: 4.228858337s proxy-resp: 0]
[2022-03-22 14:46:50.957306578 +0000 UTC m=+6318.293832110] logging.go:228: HTTP/1.1 200 POST /dbs2go-writer/bulkblocks [data: 8697 in 2 out] [remoteAddr: 188.185.121.219:35863] [X-Forwarded-For: 188.184.88.70] [X-Forwarded-Host: cmsweb-testbed.cern.ch] [auth: no-TLS cipher-none "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0192.cern.ch" wmagent@vocms0192.cern.ch X509Cert] [ref: "-" "DBSClient/Unknown/"] [req: 4.429559957s proxy-resp: 0]
[2022-03-22 15:00:17.794534604 +0000 UTC m=+7125.131060135] logging.go:228: HTTP/1.1 200 POST /dbs2go-writer/bulkblocks [data: 8042 in 2 out] [remoteAddr: 188.185.121.219:21816] [X-Forwarded-For: 188.184.88.70] [X-Forwarded-Host: cmsweb-testbed.cern.ch] [auth: no-TLS cipher-none "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0192.cern.ch" wmagent@vocms0192.cern.ch X509Cert] [ref: "-" "DBSClient/Unknown/"] [req: 2.236467875s proxy-resp: 0]
[2022-03-22 15:02:01.162415884 +0000 UTC m=+7228.498941407] logging.go:228: HTTP/1.1 200 POST /dbs2go-writer/bulkblocks [data: 8439 in 2 out] [remoteAddr: 10.100.147.128:12674] [X-Forwarded-For: 188.184.88.70] [X-Forwarded-Host: cmsweb-testbed.cern.ch] [auth: no-TLS cipher-none "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0192.cern.ch" wmagent@vocms0192.cern.ch X509Cert] [ref: "-" "DBSClient/Unknown/"] [req: 945.864915ms proxy-resp: 0]
[2022-03-22 15:02:01.162638386 +0000 UTC m=+7228.499163921] logging.go:228: HTTP/1.1 200 POST /dbs2go-writer/bulkblocks [data: 8307 in 2 out] [remoteAddr: 10.100.147.128:42517] [X-Forwarded-For: 188.184.88.70] [X-Forwarded-Host: cmsweb-testbed.cern.ch] [auth: no-TLS cipher-none "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0192.cern.ch" wmagent@vocms0192.cern.ch X509Cert] [ref: "-" "DBSClient/Unknown/"] [req: 945.571729ms proxy-resp: 0]
[2022-03-22 15:20:28.547742938 +0000 UTC m=+8335.884268449] logging.go:228: HTTP/1.1 200 POST /dbs2go-writer/bulkblocks [data: 13449 in 2 out] [remoteAddr: 10.100.6.64:7820] [X-Forwarded-For: 188.184.88.70] [X-Forwarded-Host: cmsweb-testbed.cern.ch] [auth: no-TLS cipher-none "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0192.cern.ch" wmagent@vocms0192.cern.ch X509Cert] [ref: "-" "DBSClient/Unknown/"] [req: 1.567981098s proxy-resp: 0]
```
And, this proves that integration test between WMAgent and new Go based DBSWriter is doing fine.

Alan, at this point I don't see any further need to adjust this PR and will await on your approval.

Meanwhile, I sent you and others another email on my finding with logic of `bulkblocks` API which may cause racing conditions (by design), but this is a separate story from this PR.

amaltaro commented on test/python/WMComponent_t/DBS3Buffer_t/DBSBufferFile_t.py:
```diff
@@ -947,7 +947,7 @@ def testProperties(self):
         testFileA.setDatasetPath("/Cosmics/CRUZET09-PromptReco-v1/RECO")
         testFileA.setValidStatus(validStatus="VALID")
-        testFileA.setProcessingVer(ver="ProcVer")
+        testFileA.setProcessingVer(ver="123")
```
shouldn't it be integer?

vkuznet replied:
I think we need to test code which should handle string input since there is no data-type checking anywhere in a code and moreover we know that different types can be passed to the code.

amaltaro commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -949,14 +949,25 @@ def setProcessingVersion(self, processingVersions):
 
         self.updateLFNsAndDatasets()
-        self.data.properties.processingVersion = processingVersions
+        if isinstance(processingVersions, int):
```
Did you make this change because the argument name suggests it could be multiple processing versions?
I think this is a valid case for TaskChain workflows where diff procVer can be defined for different tasks.

Unless there is a strong reason for this change, I think we should remove these changes.

vkuznet replied:
Alan, the way how this code is structured it requires a check on provided `processingVersions`. The allowed data-types (as far as I can tell) is `int, dict or list`. Therefore any other type should be ignore and proper exception should be thrown. For instance, if someone will pass the string it will be wrong since I don't know full chain of data manipulation through entire WMCore code. I still suggest to keep this change to avoid potential issue somewhere else.

amaltaro commented on src/python/WMComponent/DBS3Buffer/MySQL/UpdateDataset.py:
```diff
@@ -26,11 +26,17 @@ class UpdateDataset(DBFormatter):
+        # to be backward compatible with various places in WMCore codebase which
+        # assign processingVer to None default value. In DBS and WMCore databases
+        # the processsing version should be integer data type
+        if processingVer == None:
```
```suggestion
        if processingVer is None:
```

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSBufferFile.py:
```diff
@@ -387,8 +387,16 @@ def setProcessingVer(self, ver):
         """
-
-        self['processingVer'] = ver
+        if isinstance(ver, str) or isinstance(ver, bytes):
```
I think I commented on this code in the other PR.
When it gets to this level, processing version has been extensively tested. Maybe the only thing that we would need to keep here is something like:
```
if ver is None:
    procVer = 0
```

in case this information isn't loaded from the database.

vkuznet replied:
I disagree since python does not provide a way to check input data-type. I don't rely on your word that provided input is and will be extensively tested. By definition someone can simply pass a string to this code just calling this method. And, I don't understand why you are so resistant against checking provided data-type. I discovered exactly this issue that somewhere in WMCore someone passed a wrong data-type and I'm not convinced that changing schema will prevent from doing this again form any other place in a WMCore code. I think **IT IS SAFE** to check explicitly for correct data type and only allow the one which is expected. The code does not add anything new and only provide additional checks.

amaltaro replied:
Yes, I know it's safer. I am just trying to find a good balance between being safe and keeping code as simple as possible.

I don't think it's impossible to build a DBSBufferFile/DBSBufferBlock without relying on the DAO information, even though WMAgents do so based on DAOs. But okay, let's keep these extra checks here then. I would like to suggest a different approach though, which relies less on data type checking (I'd avoid using anything that checks for bytes/str, as it gave us a nightmare over the past years).

Here is a suggestion for this code:
``` 
ver = ver or 0
try:
    ver = int(ver)
except TypeError:
    msg = "Provided procVer=%s of type %s cannot be converted to int" % (ver, type(ver))
    raise TypeError(msg)
```

or if we do not want to "massage" the exception message, we could remove the try/except as well.
What do you think? Could you please update this code accordingly?

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py:
```diff
@@ -231,12 +231,17 @@ def setProcessingVer(self, procVer):
-        # compatibility statement for old style proc ver (still needed ?)
-        if procVer.count("-") == 1:
-            self.data["processing_era"]["processing_version"] = procVer.split("-v")[1]
+        pver = procVer
```
I think I commented on this code in the other PR.
When it gets to this level, processing version has been extensively tested. Maybe the only thing that we would need to keep here is something like:
```
if ver is None:
    procVer = 0
```

in case this information isn't loaded from the database.

vkuznet replied:
The same argument here again. It is much safer to check explicitly the provided data type and avoid potential issue with some caller which will not depend on database type definition.

amaltaro reviewed (changes requested):
@vkuznet Valentin, I left a few comments along the code.

I think you mentioned that some of those changes are needed to pass the unit tests, but it's still not clear to me why we would need them (unless the unit test is bypassing the Spec validation and testing only things from the Task object level and beyond).

Could you please make those changes I suggested/asked in a different commit? Such that we can see how jenkins is going to behave (and if we eventually find out that we need to remove these changes, we could simply remove the last commit). Thanks

vkuznet commented:
Alan, I'm not sure I understand your request to provide changes in separate commit. Please look at 6 failed jenkins tests above in this ticket which shows how tests failed. If you are interesting how the code can fail I can simply initialize the class and call `setProcessingVer` with string argument. I don't think that removing type checking make this code more generic, on a contrary it will allow the same problem which we're trying to fix here to exist in a code since by definition of Python different data-types can be passed to `setProcessingVer` as input values. The changes I propose prevent this from happening regardless of database definition, and therefore **guarantee** the correct data-type assignment to processing version.

amaltaro reviewed (changes requested):
Thanks for your comments, Valentin. I made a few more suggestions, and then I think it should be ready to go.

amaltaro reviewed (changes requested):
And I think we need one more change. Thanks for the quick turnaround.

amaltaro commented:
PS.: I am removing the link to those projects because that is meant to be done only for GH issues.

vkuznet commented:
Alan, tests are ok, but I don't know what triggered pylint warnings (even though they seem unrelated to changes I made). Please review again.

amaltaro commented:
Because pylint check works now :-D I will review this again soon.

amaltaro commented:
@vkuznet Valentin, I think there are pylint records that are worth looking into. For example:
```
src/python/WMComponent/DBS3Buffer/DBSBufferBlock.py
    W0707, line 240 in DBSBufferBlock.setProcessingVer: Consider explicitly re-raising using the 'from' keyword
    W0707, line 310 in DBSBufferBlock.setDataset: Consider explicitly re-raising using the 'from' keyword 
```

Python3 makes exceptions and traceback much more verbose. Given that you already have a proper error message implemented, when we raise an exception from within an exception block, we should actually implement it like:
```
raise TypeError(msg) from None
```

such that the exception that caused it isn't raised. You can find [https://stefan.sofa-rockers.org/2020/10/28/raise-from/](here) some very brief and clear explanation of this.

In general, if any of those **bold** pylint records have been injected by these changes, could you please revisit and clean them up?

Regarding the multiple commits, I think it pollutes the file history if we have a commit that adds something, and the next commit removes it. So, I'd be in favor of squashing both commits. For the record, [this](https://github.com/dmwm/WMCore/pull/11043/commits/37cdce4dfb95b5caf81f990c37aecdc8feeccb4c) link should remain alive and point to the WMWorkload changes that you mentioned, even after you squash and we merge it.

vkuznet commented:
ok, squashed and fixed those pylint changes which reflect the changes (but I don't touch others which pop up from previous code). Don't tell me that you don't picky :) Please review again and hopefully we can merge it.

amaltaro reviewed (approved):
It looks good to me, thanks Valentin!

You are right, there is no need to fix pylint issues from previous commits, however, if you decide to do so - even for only a few of them, like the easy and safe ones - please do so in a second commit. Saying this only for future reference (it is also in the contribution guidelines)

amaltaro commented:
For the record, you can apply this patch to WMAgents (1.5.7.patchXXX) with this command (with cmst1 user):
```
curl https://patch-diff.githubusercontent.com/raw/dmwm/WMCore/pull/11043.patch | patch -p3 -d apps/wmagentpy3/lib/python3.8/site-packages/
```

then restart the components to make sure that the new code is picked up:
```
$manage stop-agent
and
$manage start-agent
```
